fabtests/functional/flood.c

```diff
@@ -0,0 +1,319 @@
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <unistd.h>
+
+#include <shared.h>
+
+int sleep_time = 0;
+
+static ssize_t post_one_tx(struct ft_context *msg)
+{
+	ssize_t ret;
+
+	if (ft_check_opts(FT_OPT_VERIFY_DATA | FT_OPT_ACTIVE)) {
+		ret = ft_fill_buf(msg->buf + ft_tx_prefix_size(),
+				  opts.transfer_size);
+		if (ret)
+			return ret;
+	}
+
+	return ft_post_tx_buf(ep, remote_fi_addr, opts.transfer_size,
+			      NO_CQ_DATA, &msg->context, msg->buf,
+			      msg->desc, 0);
+}
+
+static ssize_t wait_check_rx_bufs(void)
+{
+	ssize_t ret;
+	int i;
+
+	ret = ft_get_rx_comp(rx_seq);
+	if (ret)
+		return ret;
+
+	if (!ft_check_opts(FT_OPT_VERIFY_DATA | FT_OPT_ACTIVE))
+		return 0;
+
+	for (i = 0; i < opts.window_size; i++) {
+		ret = ft_check_buf((char *) rx_ctx_arr[i].buf +
+				   ft_rx_prefix_size(), opts.transfer_size);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static int post_rx_sync(void)
+{
+	int ret;
+
+	ret = ft_post_rx(ep, rx_size, &rx_ctx);
+	if (ret)
+		return ret;
+
+	if (opts.dst_addr) {
+		ret = ft_tx(ep, remote_fi_addr, 1, &tx_ctx);
+		if (ret)
+			return ret;
+
+		ret = ft_get_rx_comp(rx_seq);
+	} else {
+		ret = ft_get_rx_comp(rx_seq);
+		if (ret)
+			return ret;
+
+		ret = ft_tx(ep, remote_fi_addr, 1, &tx_ctx);
+	}
+
+	return ret;
+}
+static void mr_close(struct ft_context *ctx_arr, int window_size)
+{
+	for (int i = 0; i < window_size; i++)
+	{
+		FT_CLOSE_FID(ctx_arr[i].mr);
+	}
+}
+static int run_seq_mr_send(void) {
+
+	int ret;
+
+	mr_close(tx_ctx_arr, opts.window_size);
+	mr_close(rx_ctx_arr, opts.window_size);
+
+	printf("sequential memory registration:\n");
+	ft_start();
+	if (opts.dst_addr) {
+		for (int i = 0; i < opts.window_size; i++) {
+			ret = ft_reg_mr(fi, tx_ctx_arr[i].buf, tx_mr_size,
+					ft_info_to_mr_access(fi),
+					FT_TX_MR_KEY + i, opts.iface, opts.device,
+					&(tx_ctx_arr[i].mr), &(tx_ctx_arr[i].desc));
+			if (ret)
+				return ret;
+
+			ft_post_tx_buf(ep, remote_fi_addr,  opts.transfer_size,
+				       NO_CQ_DATA,  &(tx_ctx_arr[i].context),
+				       tx_ctx_arr[i].buf, tx_ctx_arr[i].desc,
+				       ft_tag);
+			ret = ft_get_tx_comp(tx_seq);
+			if (ret)
+				return ret;
+			FT_CLOSE_FID(tx_ctx_arr[i].mr);
+		}
+	}
+	else {
+		for (int i = 0; i < opts.window_size; i++) {
+			ret = ft_reg_mr(fi, rx_ctx_arr[i].buf, rx_mr_size,
+					ft_info_to_mr_access(fi), FT_RX_MR_KEY + i,
+					opts.iface, opts.device, &(rx_ctx_arr[i].mr),
+					&(rx_ctx_arr[i].desc));
+			if (ret)
+				return ret;
+
+			ft_post_rx_buf(ep, opts.transfer_size,
+				       &(rx_ctx_arr[i].context), rx_ctx_arr[i].buf,
+				       rx_ctx_arr[i].desc, ft_tag);
+			ret = wait_check_rx_bufs();
+			if (ret)
+				return ret;
+		}
+	}
+	ft_stop();
+	if (opts.options & FT_OPT_OOB_SYNC)
+		ret = ft_sync();
+	else
+		ret = post_rx_sync();
+	if (ret)
+		return ret;
+
+	if (opts.machr)
+		show_perf_mr(opts.transfer_size, opts.window_size, &start, &end, 1,
```

Comment:
Remove the performance reporting since this is a functional test and has a hardcoded sleep to force unexpected messages. Replace with a PASS/FAIL print